rclpy/src/rclpy/wait_set.cpp

```diff
@@ -0,0 +1,296 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Include pybind11 before rclpy_common/handle.h includes Python.h
+#include <pybind11/pybind11.h>
+
+#include <rcl/error_handling.h>
+#include <rcl/rcl.h>
+#include <rcl/types.h>
+
+#include <cstring>
+#include <memory>
+#include <stdexcept>
+#include <string>
+
+extern "C"
+{
+#include "rclpy_common/common.h"
+}
+#include "rclpy_common/handle.h"
+
+#include "rclpy_common/exceptions.hpp"
+#include "wait_set.hpp"
+
+namespace rclpy
+{
+/// Destructor for a wait set
+void
+_rclpy_destroy_wait_set(PyObject * pycapsule)
+{
+  auto wait_set = static_cast<rcl_wait_set_t *>(PyCapsule_GetPointer(pycapsule, "rcl_wait_set_t"));
+
+  rcl_ret_t ret = rcl_wait_set_fini(wait_set);
+  if (ret != RCL_RET_OK) {
+    // Warning should use line number of the current stack frame
+    int stack_level = 1;
+    PyErr_WarnFormat(
+      PyExc_RuntimeWarning, stack_level,
+      "Failed to fini wait set: %s", rcl_get_error_string().str);
+    rcl_reset_error();
+  }
+
+  PyMem_Free(wait_set);
+}
+
+py::capsule
+get_zero_initialized_wait_set()
+{
+  auto deleter = [](rcl_wait_set_t * ptr) {PyMem_FREE(ptr);};
+  auto wait_set = std::unique_ptr<rcl_wait_set_t, decltype(deleter)>(
+    static_cast<rcl_wait_set_t *>(PyMem_Malloc(sizeof(rcl_wait_set_t))),
+    deleter);
+
+  *wait_set = rcl_get_zero_initialized_wait_set();
+  return py::capsule(wait_set.release(), "rcl_wait_set_t", _rclpy_destroy_wait_set);
+}
+
+void
+wait_set_init(
+  py::capsule pywait_set,
+  size_t number_of_subscriptions,
+  size_t number_of_guard_conditions,
+  size_t number_of_timers,
+  size_t number_of_clients,
+  size_t number_of_services,
+  size_t number_of_events,
+  py::capsule pycontext)
+{
+  if (0 != std::strcmp("rcl_wait_set_t", pywait_set.name())) {
+    throw py::value_error("capsule is not an rcl_wait_set_t");
+  }
+  auto wait_set = static_cast<rcl_wait_set_t *>(pywait_set);
+
+  auto context = static_cast<rcl_context_t *>(
+    rclpy_handle_get_pointer_from_capsule(pycontext.ptr(), "rcl_context_t"));
+  if (!context) {
+    throw py::error_already_set();
+  }
+
+  rcl_ret_t ret = rcl_wait_set_init(
+    wait_set,
+    number_of_subscriptions,
+    number_of_guard_conditions,
+    number_of_timers,
+    number_of_clients,
+    number_of_services,
+    number_of_events,
+    context,
+    rcl_get_default_allocator());
+  if (ret != RCL_RET_OK) {
+    throw RCLError("failed to initialize wait set");
+  }
+}
+
+void
+wait_set_clear_entities(py::capsule pywait_set)
```

Comment:
Note that this returned `True` on success before. It doesn't seem to be used, though, so this seems OK to change.